### PR TITLE
fix revision justifications and prevent unchanged responses from bein…

### PR DESCRIPTION
**Pull request by Johnetordoff**
…g submit

## QA Notes

Please make verification statements inspired by your code and what your code touches.
- Verify
- Verify

What are the areas of risk?

Any concerns/considerations/questions that development raised?

### CODE REVIEW

File: api/actions/serializers.py [hunk 1]
```diff
@@ -364,11 +364,13 @@ def create(self, validated_data):
                 raise JSONAPIAttributeException(attribute='trigger', detail='Invalid trigger.')
         except PermissionsError as exc:
             raise PermissionDenied(exc)
-        except (MachineError, ValueError):
```
Comment (Johnetordoff) [inline api/actions/serializers.py:367]: This was throwing the conflict error instead of the correct value error text.

File: api/schema_responses/views.py [hunk 2]
```diff
@@ -94,13 +94,7 @@ class SchemaResponseDetail(JSONAPIBaseView, generics.RetrieveUpdateDestroyAPIVie
     view_category = 'schema_responses'
     view_name = 'schema-responses-detail'
 
-    def get_serializer_class(self):
```
Comment (Johnetordoff) [inline api/schema_responses/views.py:97]: referencing the parent here is an SQL query that isn't committed in the transaction where the `serilaizer_class` is  AttributeError not found is thrown. It's hard to describe but in the context this error is thrown it can't make any SQL queries, so I'm just removing this function for now.

File: osf/models/schema_response.py [hunk 3]
```diff
@@ -399,7 +400,7 @@ def _validate_accept_trigger(self, user):
             # not self.pending_approvers.exists() -> called from within "approve"
             if user is None or not self.pending_approvers.exists():
                 return
-            raise ValueError(
```
Comment (Johnetordoff) [inline osf/models/schema_response.py:402]: I assume this should throw as a 409.

Comment (jwalz) [reply osf/models/schema_response.py:402]: Not technically a machine error, but work for me

File: api_tests/schema_responses/views/test_schema_response_detail.py [hunk 4]
```diff
@@ -495,6 +495,13 @@ def test_PATCH_updates_revision_justification(self, app, schema_response, payloa
         schema_response.refresh_from_db()
         assert schema_response.revision_justification == 'why not?'
 
+        payload['data']['attributes']['revision_justification'] = None
```
Comment (jwalz) [inline api_tests/schema_responses/views/test_schema_response_detail.py:498]: This should be its own test. Also, please add a test for the empty string, since I think that's what caused the issue originally.

File: osf/models/schema_response.py [hunk 5]
```diff
@@ -433,6 +434,10 @@ def _validate_reject_trigger(self, user):
 
     def _on_submit(self, event_data):
         '''Add the provided approvers to pending_approvers and set the submitted_timestamp.'''
+        if not self.updated_response_keys:
```
Comment (jwalz) [inline osf/models/schema_response.py:437]: `if not self.updated_response_keys or not self.revision_justification:`

File: api_tests/actions/views/test_schema_response_action_list.py [hunk 6]
```diff
@@ -391,6 +391,19 @@ def test_POST_status_code__moderator_reject(self, app, registration_status, role
 @pytest.mark.enable_quickfiles_creation
 class TestSchemaResponseActionListPOSTBehavior:
 
+    def test_POST_submit__denies_unchanged_submission(self, app):
```
Comment (jwalz) [inline api_tests/actions/views/test_schema_response_action_list.py:394]: add test for submit without a justification, too

File: api_tests/schema_responses/views/test_schema_response_detail.py [hunk 7]
```diff
@@ -495,6 +495,13 @@ def test_PATCH_updates_revision_justification(self, app, schema_response, payloa
         schema_response.refresh_from_db()
         assert schema_response.revision_justification == 'why not?'
 
+    def test_PATCH_empty_revision_justification_fails(self, app, schema_response, payload, admin_user):
```
Comment (jwalz) [inline api_tests/schema_responses/views/test_schema_response_detail.py:498]: `test_PATCH. . ._fails`?

Comment (Johnetordoff) [reply api_tests/schema_responses/views/test_schema_response_detail.py:498]: ahh forgot to rename it when I reverted the changes

File: api_tests/actions/views/test_schema_response_action_list.py [hunk 8]
```diff
@@ -68,6 +68,13 @@ def configure_test_preconditions(
         schema_response.pending_approvers.add(
             *[user for user, _ in registration.get_admin_contributors_recursive()]
         )
+
+    if schema_response_state is ApprovalStates.IN_PROGRESS:
+        # need valid changes for submission validations
+        schema_response.update_responses({'q1': 'update for submission'})
```
Comment (jwalz) [inline api_tests/actions/views/test_schema_response_action_list.py:74]: we're setting the responses above (L62), so this shouldn't be needed, but it's also not breaking anything, so is fine for the moment

**Review by jwalz — APPROVED** [review]
